channeldb/migration26/migration.go

```diff
@@ -0,0 +1,498 @@
+	openChanBucket := tx.ReadWriteBucket(openChannelBucket)
+
+	// If no bucket is found, we can exit early.
+	if openChanBucket == nil {
+		return false, fmt.Errorf("root bucket not found")
+	}
+
+	// Locate the next migration height.
+	locator, err := locateNextUpdateNum(openChanBucket)
+	if err != nil {
+		return false, fmt.Errorf("locator got error: %v", err)
+	}
+
+	// If the returned locator is nil, we've done migrating the logs.
+	if locator == nil {
+		return true, nil
+	}
+
+	// Read a list of old revocation logs.
+	entryMap, err := readOldRevocationLogs(openChanBucket, locator)
+	if err != nil {
+		return false, fmt.Errorf("read old logs err: %v", err)
+	}
+
+	// Migrate the revocation logs.
+	return false, writeRevocationLogs(openChanBucket, entryMap)
+}
+
+// deleteOldBuckets iterates all the channel buckets and deletes the old
+// revocation buckets.
+func deleteOldBuckets(db kvdb.Backend) error {
+	// locators records all the chan buckets found in the database.
+	locators := make([]*updateLocator, 0)
+
+	// reader is a helper closure that saves the locator found. Each
+	// locator is relatively small(33+32+36+8=109 bytes), assuming 1 GB of
+	// ram we can fit roughly 10 million records. Since each record
+	// corresponds to a channel, we should have more than enough memory to
+	// read them all.
+	reader := func(_ kvdb.RwBucket, l *updateLocator) error { // nolint:unparam
+		locators = append(locators, l)
+		return nil
+	}
+
+	// remover is a helper closure that removes the old revocation log
+	// bucket under the specified chan bucket by the given locator.
+	remover := func(rootBucket kvdb.RwBucket, l *updateLocator) error {
+		chanBucket, err := l.locateChanBucket(rootBucket)
+		if err != nil {
+			return err
+		}
+
+		return chanBucket.DeleteNestedBucket(
+			revocationLogBucketDeprecated,
+		)
+	}
+
+	// Perform the deletion in one db transaction. This should not cause
+	// any memory issue as the deletion doesn't load any data from the
+	// buckets.
+	return kvdb.Update(db, func(tx kvdb.RwTx) error {
+		openChanBucket := tx.ReadWriteBucket(openChannelBucket)
+
+		// Exit early if there's no bucket.
+		if openChanBucket == nil {
+			return nil
+		}
+
+		// Iterate the buckets to find all the locators.
+		err := iterateBuckets(openChanBucket, nil, reader)
+		if err != nil {
+			return err
+		}
+
+		// Iterate the locators and delete all the old revocation log
+		// buckets.
+		for _, l := range locators {
+			err := remover(openChanBucket, l)
+			// If the bucket doesn't exist, we can exit safety.
+			if err != nil && err != kvdb.ErrBucketNotFound {
+				return err
+			}
+		}
+
+		return nil
+	}, func() {})
+}
+
+// writeRevocationLogs unwraps the entryMap and writes the new revocation logs.
+func writeRevocationLogs(openChanBucket kvdb.RwBucket,
+	entryMap logEntries) error {
+
+	for locator, logs := range entryMap {
+		// Find the channel bucket.
+		chanBucket, err := locator.locateChanBucket(openChanBucket)
```

Comment:
Pretty sure the locator's height needs to be copied, the cursor returns read-only values directly from bbolt's mmap, which can get resized here since mutations (insertions) are happening here. Using old mmap values either gives corrupt values or a panic

Reply:
We don't use the locator's height here, maybe I should've separated the two types of locators.

Reply:
What's used here are `nodePub`, `chainhash` and `fundingOutpoint`. Since we are not touching any of the upper-level buckets, and only working on the lowest-nested bucket that's located by these three keys, I'm assuming it's fine as the insertions happen in a new block of memory/disk space?

Also I think the `updateLocator`'s pointer is not managed by bbolt and, when we create it, we just copy the values read from that cursor.

Meanwhile I agree that it's better to use `updateLocator` instead of `*updateLocator` everywhere to avoid concerns like such.